mmtk/src/gc_trigger.rs

```diff
@@ -0,0 +1,330 @@
+    /// We will need to take them into consideration so that the new heap size can accomodate those allocations.
+    pending_pages: AtomicUsize,
+}
+
+impl JuliaGCTrigger {
+    pub fn new(max_total_mem: usize) -> Self {
+        // values as defined in gc-stock.c
+        Self {
+            heap_target: AtomicUsize::new(DEFAULT_COLLECT_INTERVAL),
+            max_total_memory: AtomicUsize::new(max_total_mem),
+            old_pause_time: AtomicUsize::new(1e7 as usize),
+            old_mut_time: AtomicUsize::new(1e9 as usize),
+            old_heap_size: AtomicUsize::new(0),
+            old_alloc_diff: AtomicUsize::new(DEFAULT_COLLECT_INTERVAL),
+            old_freed_diff: AtomicUsize::new(DEFAULT_COLLECT_INTERVAL),
+            gc_start_time: AtomicUsize::new(0),
+            gc_end_time: AtomicUsize::new(0),
+            mutator_time: AtomicUsize::new(0),
+            thrash_counter: AtomicUsize::new(0),
+            thrashing: AtomicUsize::new(0),
+            before_free_heap_size: AtomicUsize::new(0),
+            prev_sweep_full: AtomicBool::new(true),
+            pending_pages: AtomicUsize::new(0),
+        }
+    }
+}
+
+impl GCTriggerPolicy<JuliaVM> for JuliaGCTrigger {
+    fn on_gc_start(&self, mmtk: &'static MMTK<JuliaVM>) {
+        // reserved pages now should include pending allocations
+        let reserved_pages_now =
+            mmtk.get_plan().get_reserved_pages() + self.pending_pages.load(Ordering::SeqCst);
+
+        let now = unsafe { jl_hrtime() };
+        self.gc_start_time.store(now as usize, Ordering::Relaxed);
+
+        self.mutator_time.store(
+            if self.gc_end_time.load(Ordering::Relaxed) == 0 {
+                self.old_mut_time.load(Ordering::Relaxed)
+            } else {
+                self.gc_start_time.load(Ordering::Relaxed)
+                    - self.gc_end_time.load(Ordering::Relaxed)
+            },
+            Ordering::Relaxed,
+        );
+        self.before_free_heap_size.store(
+            conversions::pages_to_bytes(reserved_pages_now),
+            Ordering::Relaxed,
+        );
+
+        self.prev_sweep_full.store(
+            if let Some(gen) = mmtk.get_plan().generational() {
+                gen.last_collection_full_heap()
+            } else {
+                false
+            },
+            Ordering::Relaxed,
+        );
+        info!(
+            "On GC start: \
+            reserved_pages_now = {}, \
+            gc_start_time = {}, \
+            mutator_time = {}, \
+            before_free_heap_size = {}, \
+            prev_sweep_full = {}",
+            reserved_pages_now,
+            self.gc_start_time.load(Ordering::Relaxed),
+            self.mutator_time.load(Ordering::Relaxed),
+            self.before_free_heap_size.load(Ordering::Relaxed),
+            self.prev_sweep_full.load(Ordering::Relaxed),
+        );
+    }
+    fn on_gc_end(&self, mmtk: &'static MMTK<JuliaVM>) {
+        // FIXME: ignoring prev_sweep_full until we implement stickyImmix
```

Comment:
We may forget this when we add sticky immix. If possible, we can just add the code for generational GCs, though they will not be executed.